core/network/impl/state_sync_request_flow.cpp

```diff
@@ -6,118 +6,145 @@
 #include <boost/algorithm/string/predicate.hpp>
 
 #include "network/impl/state_sync_request_flow.hpp"
-#include "runtime/common/uncompress_code_if_needed.hpp"
-#include "runtime/module.hpp"
-#include "runtime/module_factory.hpp"
-#include "runtime/runtime_api/impl/core.hpp"
-#include "runtime/runtime_context.hpp"
 #include "storage/predefined_keys.hpp"
-#include "storage/trie/serialization/trie_serializer.hpp"
-#include "storage/trie_pruner/trie_pruner.hpp"
-
-OUTCOME_CPP_DEFINE_CATEGORY(kagome::network, StateSyncRequestFlow::Error, e) {
-  using E = decltype(e);
-  switch (e) {
-    case E::EMPTY_RESPONSE:
-      return "State sync empty response";
-    case E::HASH_MISMATCH:
-      return "State sync hash mismatch";
-  }
-  abort();
-}
+#include "storage/trie/compact_decode.hpp"
+#include "storage/trie/serialization/polkadot_codec.hpp"
+#include "storage/trie/trie_storage_backend.hpp"
 
 namespace kagome::network {
   StateSyncRequestFlow::StateSyncRequestFlow(
-      std::shared_ptr<storage::trie_pruner::TriePruner> state_pruner,
+      std::shared_ptr<storage::trie::TrieStorageBackend> db,
       const primitives::BlockInfo &block_info,
       const primitives::BlockHeader &block)
-      : state_pruner_{state_pruner}, block_info_{block_info}, block_{block} {
-    BOOST_ASSERT(state_pruner_ != nullptr);
-    roots_.insert({std::nullopt, Root{}});
+      : db_{std::move(db)},
+        block_info_{block_info},
+        block_{block},
+        log_{log::createLogger("StateSync")} {
+    done_ = isKnown(block.state_root);
+    if (not done_) {
+      auto &level = levels_.emplace_back();
+      level.child = {};
+      level.branch_hash = block.state_root;
+    }
   }
 
   bool StateSyncRequestFlow::complete() const {
-    return complete_roots_.size() == roots_.size();
+    return done_;
   }
 
   StateRequest StateSyncRequestFlow::nextRequest() const {
     BOOST_ASSERT(not complete());
-    return StateRequest{block_info_.hash, last_key_, true};
+    StateRequest req{block_info_.hash, {}, false};
+    for (auto &level : levels_) {
+      storage::trie::KeyNibbles nibbles;
+      for (auto &item : level.stack) {
+        nibbles.put(item.node->getKeyNibbles());
+        if (item.branch) {
+          nibbles.putUint8(*item.branch);
+        }
+      }
+      if (nibbles.size() % 2 != 0) {
+        nibbles.putUint8(0);
+      }
+      req.start.emplace_back(nibbles.toByteBuffer());
+    }
+    return req;
   }
 
   outcome::result<void> StateSyncRequestFlow::onResponse(
       const StateResponse &res) {
+    storage::trie::PolkadotCodec codec;
```

Comment:
would it make sense to make this codec as static variable, not to initialize new object every time?

Reply:
That's not class, just static methods